x/batching/keeper/abci.go

```diff
@@ -0,0 +1,161 @@
+package keeper
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+
+	"cosmossdk.io/collections"
+	"cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/sedaprotocol/seda-chain/cmd/sedad/utils"
+	"github.com/sedaprotocol/seda-chain/x/batching/types"
+	tallytypes "github.com/sedaprotocol/seda-chain/x/tally/types"
+)
+
+func (k Keeper) EndBlock(ctx sdk.Context) (err error) {
+	// Use defer to prevent returning an error, which would cause
+	// the chain to halt.
+	defer func() {
+		// Handle a panic.
+		if r := recover(); r != nil {
+			k.Logger(ctx).Error("recovered from panic in batching EndBlock", "err", r)
+		}
+		// Handle an error.
+		if err != nil {
+			k.Logger(ctx).Error("error in batching EndBlock", "err", err)
+		}
+		err = nil
+	}()
+
+	batch, err := k.ConstructBatch(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = k.SetBatch(ctx, batch)
+	if err != nil {
+		return err
+	}
+	err = k.IncrementCurrentBatchNum(ctx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (k Keeper) ConstructBatch(ctx sdk.Context) (types.Batch, error) {
+	curBatchNum, err := k.GetCurrentBatchNum(ctx)
+	if err != nil {
+		return types.Batch{}, err
+	}
+	dataLeaves, dataRootHex, err := k.ConstructDataResultTree(ctx)
+	if err != nil {
+		return types.Batch{}, err
+	}
+	valLeaves, valRootHex, err := k.ConstructValidatorTree(ctx)
+	if err != nil {
+		return types.Batch{}, err
+	}
+
+	return types.Batch{
+		BatchNumber:      curBatchNum,
+		BlockHeight:      ctx.BlockHeight(),
+		DataResultRoot:   dataRootHex,
+		ValidatorRoot:    valRootHex,
+		DataResultLeaves: dataLeaves,
+		ValidatorLeaves:  valLeaves,
+		BlockTime:        ctx.BlockTime(),
+	}, nil
+}
+
+// ConstructDataResultTree constructs a data result tree based on the
+// batching-ready data results returned from the core contract and
+// returns its leaves and its hex-encoded tree root.
+func (k Keeper) ConstructDataResultTree(ctx sdk.Context) ([][]byte, string, error) {
+	coreContract, err := k.wasmStorageKeeper.GetCoreContractAddr(ctx)
+	if err != nil {
+		return nil, "", err
+	}
+	// TODO: Deal with offset and limits. (#313)
+	queryRes, err := k.wasmViewKeeper.QuerySmart(ctx, coreContract, []byte(`{"get_data_results_by_status":{"status": "tallied", "offset": 0, "limit": 100}}`))
```

Comment:
wonder if would be safer/better to have this as a struct instead of raw JSON

Reply:
It's mostly just to avoid marshaling it every time we use it..